src/uct/cuda/cuda_copy/cuda_copy_iface.h

```diff
@@ -8,25 +8,76 @@
 
 #include <uct/base/uct_iface.h>
 #include <uct/cuda/base/cuda_iface.h>
+#include <ucs/memory/memory_type.h>
 #include <pthread.h>
 
 
-typedef uint64_t uct_cuda_copy_iface_addr_t;
+typedef enum uct_cuda_copy_loc {
+    UCT_CUDA_COPY_LOC_HOST,
+    UCT_CUDA_COPY_LOC_PINNED,
+    UCT_CUDA_COPY_LOC_MANAGED,
+    UCT_CUDA_COPY_LOC_LAST
+} uct_cuda_copy_loc_t;
+
+
+#define uct_cuda_copy_get_mem_type_index(mem_type, index) { \
+    switch (mem_type) { \
+        case UCS_MEMORY_TYPE_HOST: \
+            index = UCT_CUDA_COPY_LOC_HOST; \
+            break; \
+        case UCS_MEMORY_TYPE_CUDA: \
+            index = UCT_CUDA_COPY_LOC_PINNED; \
+            break; \
+        case UCS_MEMORY_TYPE_CUDA_MANAGED: \
+            index = UCT_CUDA_COPY_LOC_MANAGED; \
```

Comment:
will help to use the different stream for cuda and managed in the same direction?

Reply:
Prevents false serialization.

Reply:
how many copy engines (CE) we have from each direction in the hardware?   isn't it driver serialize H2D on same context?
isn't it half of PCI bandwidth if we do two concurrent H2D transfers?

Reply:
CEs will likely vary by generation; and serialization can occur if PCI bandwidth is full saturated. Having separate streams will ensure that two independent transfers (one which has pinned device as source and another which has managed memory as source) doesn't unnecessarily depend on each other by being queued on the same stream. This way a D2H transfer that saturates half pcie bandwidth will not lock out another concurrent M2H transfer (which on its own saturates half the bandwidth) from progressing concurrently by being enqueued on the same stream and waiting for the previous stream operation (the D2H operation) to complete. This will allow full saturation for the specific example. Instead if we re-use the same stream, then we will guarantee that at best only the saturation is reached.

Reply:
do we have use case and data which shows  false serialization because of current streams?  Otherwise, IMHO we are making this complex logic which i feel not necessary now. And also additional cuda mem queries in for each copy.  Even in case of using CUDA and MANAGED memory simultaneously,  cuda should be using GDR and managed memory might be in pipeline so they both wont be using cuda_copy simultaneously in optimal case.  with GDR off, we wanted to complete outstanding pipeline fragments fast in the order they submitted.

IMHO, we can simply go with cudaMemcpyDefault for handling data direction and memtypes  automatically.

Reply:
> do we have use case and data which shows false serialization because of current streams?

False serialization can easily occur even without managed memory use. We're adding streams for each data path in this PR. Consider an alltoall where all the data goes through staging. Intra-node device-to-device may need to go through device buffers while inter-node device-to-device may need to go through host buffers (because GDR isn't available). If we don't add a stream for D2D path (for user to bounce buffer part of the pipeline), we essentially serialize the inter-node and inter-node transfers because we're forcing both operations onto the same stream. Alltoall being common enough, we're likely to hit this issue on non-GDR platforms.

> Otherwise, IMHO we are making this complex logic which i feel not necessary now. 
> IMHO, we can simply go with cudaMemcpyDefault for handling data direction and memtypes automatically.

Don't really see this as complex logic as we're simply adding a stream for each data path. The overhead of querying for pointer type in the large message region is negligible which may even go away with UCTs being able to query memory types through a cache going forward.

Reply:
@bureddy If you point to how this PR introduces not insignificant performance overheads or to specific design limitations in using dedicated streams for each data path and drawbacks for concurrent transfer case, I'm happy to make changes. Please let me know.